src/audio/mux/mux.c

```diff
@@ -200,7 +205,231 @@ static struct comp_dev *mux_new(const struct comp_driver *drv,
+	assert(!ret);
+	ret = build_config(cd, dev);
+
+	if (ret < 0) {
+		rfree(cd);
+		rfree(dev);
+		return NULL;
+	}
+
+	dev->state = COMP_STATE_READY;
+	return dev;
+}
+
+/* In ipc4 case param is figured out by module config so we need to first
+ * set up param then verify param. BTW for IPC3 path, the param is sent by
+ * host driver.
+ */
+static void set_mux_params(struct comp_dev *dev,
+			   struct sof_ipc_stream_params *params)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer *sink, *source;
+	struct comp_buffer __sparse_cache *sink_c, *source_c;
+	struct list_item *source_list;
+	int i, j, valid_bit_depth;
+	const uint32_t byte_align = 1;
+	const uint32_t frame_align_req = 1;
+
+	params->direction = dev->direction;
+	params->channels =  cd->md.base_cfg.audio_fmt.channels_count;
+	params->rate = cd->md.base_cfg.audio_fmt.sampling_frequency;
+	params->sample_container_bytes = cd->md.base_cfg.audio_fmt.depth / 8;
+	params->sample_valid_bytes = cd->md.base_cfg.audio_fmt.valid_bit_depth / 8;
+	params->buffer_fmt = cd->md.base_cfg.audio_fmt.interleaving_style;
+	params->buffer.size = cd->md.base_cfg.ibs;
+	params->no_stream_position = 1;
+
+	/* There are two input pins and one output pin in the mux.
+	 * For the first input we assign parameters from base_cfg,
+	 * for the second from reference_format
+	 * and for sink output_format.
+	 */
+
+	/* update sink format */
+	if (!list_is_empty(&dev->bsink_list)) {
+		sink = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+		sink_c = buffer_acquire(sink);
+		audio_stream_init_alignment_constants(byte_align, frame_align_req,
+						      &sink_c->stream);
+
+		if (!sink_c->hw_params_configured) {
+			struct ipc4_audio_format out_fmt;
+
+			out_fmt = cd->md.output_format;
+			sink_c->stream.channels = out_fmt.channels_count;
+			sink_c->stream.rate = out_fmt.sampling_frequency;
+			audio_stream_fmt_conversion(out_fmt.depth,
+						    out_fmt.valid_bit_depth,
+						    &sink_c->stream.frame_fmt,
+						    &sink_c->stream.valid_sample_fmt,
+						    out_fmt.s_type);
+
+			sink_c->buffer_fmt = out_fmt.interleaving_style;
+			params->frame_fmt = sink_c->stream.frame_fmt;
+
+			for (i = 0; i < SOF_IPC_MAX_CHANNELS; i++)
+				sink_c->chmap[i] = (out_fmt.ch_map >> i * 4) & 0xf;
+
+			sink_c->hw_params_configured = true;
+		}
+		buffer_release(sink_c);
+	}
+
+	/* update each source format */
+	if (!list_is_empty(&dev->bsource_list)) {
+		list_for_item(source_list, &dev->bsource_list)
+		{
+			source = container_of(source_list, struct comp_buffer, sink_list);
+			source_c = buffer_acquire(source);
+			audio_stream_init_alignment_constants(byte_align, frame_align_req,
+							      &source_c->stream);
+			j = source_c->id;
+			cd->config.streams[j].pipeline_id = source_c->pipeline_id;
+			valid_bit_depth = cd->md.base_cfg.audio_fmt.valid_bit_depth;
+			/* set parameters for speech input channels */
+			if (j == BASEFW_QUEUE_ID) {
```

Comment:
Sorry my bad, I guess this has nothing to do with BASEFW, so better to rename this to BASE_CFG_QUEUED_ID .

And you still have the confusing comment about "speech input channels", please address this (either expand why this is related to speech, or replace with something understandable).